lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+                // Merge two registers into a single argument.
+                size_t Size = model::Register::getSize(UpperRegister)
+                              + model::Register::getSize(LowerRegister);
+                Upper.Type = ensurePrimitiveTypeIsAvailable(UpperKind,
+                                                            Size,
+                                                            Binary,
+                                                            TypeDependencies);
+
+                // Drop the merged half of the current type from the list.
+                Result.pop_back();
+
+                WasHandledSpecifically = true;
+              }
+            }
           }
+        }
 
-          Result.pop_back();
-        } else {
-          return std::nullopt;
+        if (!WasHandledSpecifically) {
+          // TODO: It might still be worth it to add this unused register as
+          // an argument to indicate the register ordering.
         }
+      } else {
+        // Ignore this register, it's most like unused.
       }
 
+      bool IsUsed = (CurrentRegisterIterator != Used.end());
       MustUseTheNextOne = MustUseTheNextOne || IsUsed;
     }
 
+    // Ensure all the indices are consistent.
     for (auto Pair : llvm::enumerate(llvm::reverse(Result)))
       Pair.value().Index = Pair.index();
 
     return Result;
   }
 
-  static llvm::SmallVector<model::Argument, 8>
-  convertStackArguments(model::QualifiedType StackArgumentTypes,
-                        size_t IndexOffset) {
+  /// Helper used for converting stack argument struct into
+  /// the c-style representation
+  ///
+  /// \param StackArgumentTypes The qualified type of the relevant part of
+  ///        the stack.
+  /// \param IndexOffset The index of the first argument (indicates the register
+  ///        argument count).
+  ///
+  /// \return An ordered list of arguments.
+  static std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                              size_t IndexOffset) {
+    // Qualifiers here are not allowed since this must point to a simple struct.
     revng_assert(StackArgumentTypes.Qualifiers.empty());
+
     auto *Unqualified = StackArgumentTypes.UnqualifiedType.get();
-    if (not Unqualified)
-      return {};
+    if (!Unqualified) {
+      // If there is no type, it means that the importer responsible for this
+      // function didn't detect any stack arguments and avoided creating
+      // a new empty type.
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
 
     auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
     revng_assert(Pointer != nullptr,
                  "`RawFunctionType::StackArgumentsType` must be a struct");
     const model::StructType &Types = *Pointer;
 
+    // If the struct is empty, there's no arguments to extract from it.
+    if (Types.Fields.empty()) {
+      revng_assert(Types.size() == 0);
```

Comment:
This assumption is not guaranteed to hold, we can't assert.

If you convert back to `RawFunctionType`, the stack will disappear, which corrupts the semantic.
I think we should introduce a new concept: in order to be successful, the conversion from `Raw` to `CABI` must *consume* all of the stack.

In practice we must be able to say that the size of the stack arguments does not change if we do `toRaw(toCABI(Raw))`.

An alternative would be to forge a new `struct` and add it as an argument to the resulting `CABIFunctionType`, but I think for now it's better to just fail the conversion, which is a legitimate outcome.

Anyway, given that we plan to run `toCABI` post-DLA, this shouldn't happen.

Reply:
Ah, wait, yeah, thanks for spotting that!

Luckily this couldn't have resulted in anything bad since if it runs into a non-0-sized struct with no fields, it would just assert.

So, which option do you prefer, to outright fail if there's a field missing in the struct where one should be (this is consistent with non-trivially aligned structs) or should we just "fill" the missing spots with pointer-sized fields and be done with it? I'm double-checking because you seem to lean to the first approach where-as I kind of prefer the second one - since IMO, the more functions we can potentially represent as CABI-FT, the less the chance that some of our users would need to deal with RawFTs.

Reply:
> the more functions we can potentially represent as CABI-FT, the less the chance that some of our users would need to deal with RawFTs

The ABI conversion is complex, I want to follow this approach: let's first support what we expect given the typical pipeline we run (i.e., we assume DLA has been run) and keep it simple, then we look at the results and if we see bad things we fix them specifically.

Trying to handle the general case increases complexity to handle situations that will never pop up and increases the latency to merging.
Let's focus on what we expect should work, test it and then improve incrementally.

By design this conversion is 1) best effort, 2) tuned to prettify the results we typically get from `DetectABI` and `DLA`. Let's stick to it.

tl;dr: bail out and fix later if it happens (it shouldn't).

Reply:
Sounds good.

Reply:
Forgot to close this one: it is also gone.